Not possible to unset the extra_filter_expression

If you have a perfectly valid `filter_object` it should be possible to unset the `extra_filter_expression`. 

Upon creation, you actually *can* create a recipe with only a `filter_object`. You just have to NOT set `extra_filter_expression` in the JSON data you send. 

I threw in this test on v2 to check:
```python
        def test_it_can_update_recipes_with_only_filter_object(self, api_client):
            recipe = RecipeFactory(name="unchanged", extra_filter_expression="true")

            res = api_client.patch(
                "/api/v2/recipe/%s/" % recipe.id,
                {"name": "changed", "extra_filter_expression": "", "filter_object": [
                    {"type": "channel", "channels": ["beta", "release"]}
                ]},
            )
            assert res.status_code == 200, res.json()
```
It fails with:
```
>       assert res.status_code == 200, res.json()
E       AssertionError: {'extra_filter_expression': ['This field may not be blank.']}
E       assert 400 == 200
E        +  where 400 = <Response status_code=400, "application/json">.status_code
```

--- a/normandy/recipes/api/v2/serializers.py
+++ b/normandy/recipes/api/v2/serializers.py
@@ -64,7 +64,7 @@ class RecipeSerializer(serializers.ModelSerializer):
     approved_revision = RecipeRevisionSerializer(read_only=True)
     arguments = serializers.JSONField()
     enabled = serializers.BooleanField(read_only=True)
-    extra_filter_expression = serializers.CharField(required=False)
+    extra_filter_expression = serializers.CharField(required=False, allow_blank=True)
     filter_expression = serializers.CharField(read_only=True)
     filter_object = serializers.JSONField(required=False)
     last_updated = serializers.DateTimeField(read_only=True)
@@ -127,21 +127,22 @@ class RecipeSerializer(serializers.ModelSerializer):
         return self.update(recipe, validated_data)
 
     def validate_extra_filter_expression(self, value):
-        jexl = JEXL()
+        if value:
+            jexl = JEXL()
 
-        # Add mock transforms for validation. See
-        # http://normandy.readthedocs.io/en/latest/user/filter_expressions.html#transforms
-        # for a list of what transforms we expect to be available.
-        jexl.add_transform("date", lambda x: x)
-        jexl.add_transform("stableSample", lambda x: x)
-        jexl.add_transform("bucketSample", lambda x: x)
-        jexl.add_transform("preferenceValue", lambda x: x)
-        jexl.add_transform("preferenceIsUserSet", lambda x: x)
-        jexl.add_transform("preferenceExists", lambda x: x)
+            # Add mock transforms for validation. See
+            # http://normandy.readthedocs.io/en/latest/user/filter_expressions.html#transforms
+            # for a list of what transforms we expect to be available.
+            jexl.add_transform("date", lambda x: x)
+            jexl.add_transform("stableSample", lambda x: x)
+            jexl.add_transform("bucketSample", lambda x: x)
+            jexl.add_transform("preferenceValue", lambda x: x)
+            jexl.add_transform("preferenceIsUserSet", lambda x: x)
+            jexl.add_transform("preferenceExists", lambda x: x)
 
-        errors = list(jexl.validate(value))
-        if errors:
-            raise serializers.ValidationError(errors)
+            errors = list(jexl.validate(value))
+            if errors:
+                raise serializers.ValidationError(errors)
 
         return value
 
@@ -207,6 +208,16 @@ class RecipeSerializer(serializers.ModelSerializer):
                         "one of extra_filter_expression or filter_object is required"
                     )
                 elif len(data["filter_object"]) == 0:
+                    raise serializers.ValidationError(
+                        "if extra_filter_expression is blank, "
+                        "at least one filter_object is required"
+                    )
+        else:
+            if "extra_filter_expression" in data or "filter_object" in data:
+                # If either is attempted to be updated, at least one of them must be truthy.
+                if not data.get("extra_filter_expression", "").strip() and not data.get(
+                    "filter_object"
+                ):
                     raise serializers.ValidationError(
                         "if extra_filter_expression is blank, "
                         "at least one filter_object is required"

--- a/normandy/recipes/api/v2/serializers.py
+++ b/normandy/recipes/api/v2/serializers.py
@@ -203,14 +203,9 @@ class RecipeSerializer(serializers.ModelSerializer):
 
         if self.instance is None:
             if data.get("extra_filter_expression", "").strip() == "":
-                if "filter_object" not in data:
+                if not data.get("filter_object"):
                     raise serializers.ValidationError(
                         "one of extra_filter_expression or filter_object is required"
-                    )
-                elif len(data["filter_object"]) == 0:
-                    raise serializers.ValidationError(
-                        "if extra_filter_expression is blank, "
-                        "at least one filter_object is required"
                     )
         else:
             if "extra_filter_expression" in data or "filter_object" in data:

--- a/normandy/recipes/api/v3/serializers.py
+++ b/normandy/recipes/api/v3/serializers.py
@@ -207,14 +207,9 @@ class RecipeSerializer(serializers.ModelSerializer):
 
         if self.instance is None:
             if data.get("extra_filter_expression", "").strip() == "":
-                if "filter_object" not in data:
+                if not data.get("filter_object"):
                     raise serializers.ValidationError(
                         "one of extra_filter_expression or filter_object is required"
-                    )
-                elif len(data["filter_object"]) == 0:
-                    raise serializers.ValidationError(
-                        "if extra_filter_expression is blank, "
-                        "at least one filter_object is required"
                     )
         else:
             if "extra_filter_expression" in data or "filter_object" in data:

--- a/normandy/recipes/api/v3/serializers.py
+++ b/normandy/recipes/api/v3/serializers.py
@@ -64,7 +64,7 @@ class RecipeSerializer(serializers.ModelSerializer):
     approved_revision = RecipeRevisionSerializer(read_only=True)
     arguments = serializers.JSONField()
     enabled = serializers.BooleanField(read_only=True)
-    extra_filter_expression = serializers.CharField(required=False)
+    extra_filter_expression = serializers.CharField(required=False, allow_blank=True)
     filter_expression = serializers.CharField(read_only=True)
     filter_object = serializers.JSONField(required=False)
     last_updated = serializers.DateTimeField(read_only=True)
@@ -131,21 +131,22 @@ class RecipeSerializer(serializers.ModelSerializer):
         return self.update(recipe, validated_data)
 
     def validate_extra_filter_expression(self, value):
-        jexl = JEXL()
+        if value:
+            jexl = JEXL()
 
-        # Add mock transforms for validation. See
-        # http://normandy.readthedocs.io/en/latest/user/filter_expressions.html#transforms
-        # for a list of what transforms we expect to be available.
-        jexl.add_transform("date", lambda x: x)
-        jexl.add_transform("stableSample", lambda x: x)
-        jexl.add_transform("bucketSample", lambda x: x)
-        jexl.add_transform("preferenceValue", lambda x: x)
-        jexl.add_transform("preferenceIsUserSet", lambda x: x)
-        jexl.add_transform("preferenceExists", lambda x: x)
+            # Add mock transforms for validation. See
+            # http://normandy.readthedocs.io/en/latest/user/filter_expressions.html#transforms
+            # for a list of what transforms we expect to be available.
+            jexl.add_transform("date", lambda x: x)
+            jexl.add_transform("stableSample", lambda x: x)
+            jexl.add_transform("bucketSample", lambda x: x)
+            jexl.add_transform("preferenceValue", lambda x: x)
+            jexl.add_transform("preferenceIsUserSet", lambda x: x)
+            jexl.add_transform("preferenceExists", lambda x: x)
 
-        errors = list(jexl.validate(value))
-        if errors:
-            raise serializers.ValidationError(errors)
+            errors = list(jexl.validate(value))
+            if errors:
+                raise serializers.ValidationError(errors)
 
         return value
 
@@ -211,6 +212,16 @@ class RecipeSerializer(serializers.ModelSerializer):
                         "one of extra_filter_expression or filter_object is required"
                     )
                 elif len(data["filter_object"]) == 0:
+                    raise serializers.ValidationError(
+                        "if extra_filter_expression is blank, "
+                        "at least one filter_object is required"
+                    )
+        else:
+            if "extra_filter_expression" in data or "filter_object" in data:
+                # If either is attempted to be updated, at least one of them must be truthy.
+                if not data.get("extra_filter_expression", "").strip() and not data.get(
+                    "filter_object"
+                ):
                     raise serializers.ValidationError(
                         "if extra_filter_expression is blank, "
                         "at least one filter_object is required"

--- a/normandy/recipes/tests/api/v2/test_api.py
+++ b/normandy/recipes/tests/api/v2/test_api.py
@@ -181,6 +181,23 @@ class TestRecipeAPI(object):
         def test_it_can_create_extra_filter_expression_omitted(self, api_client):
             action = ActionFactory()
 
+            # First try to create a recipe with 0 filter objects.
+            res = api_client.post(
+                "/api/v2/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "filter_object": [],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 400
+            assert res.json()["non_field_errors"] == [
+                "one of extra_filter_expression or filter_object is required"
+            ]
+
+            # Setting at least some filter_object but omitting the extra_filter_expression.
             res = api_client.post(
                 "/api/v2/recipe/",
                 {

--- a/normandy/recipes/tests/api/v2/test_api.py
+++ b/normandy/recipes/tests/api/v2/test_api.py
@@ -157,6 +157,47 @@ class TestRecipeAPI(object):
             recipes = Recipe.objects.all()
             assert recipes.count() == 1
 
+        def test_it_can_create_recipes_with_only_filter_object(self, api_client):
+            action = ActionFactory()
+
+            res = api_client.post(
+                "/api/v2/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "extra_filter_expression": "   ",
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 201, res.json()
+
+            assert Recipe.objects.count() == 1
+            recipe = Recipe.objects.get()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
+        def test_it_can_create_extra_filter_expression_omitted(self, api_client):
+            action = ActionFactory()
+
+            res = api_client.post(
+                "/api/v2/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 201, res.json()
+
+            assert Recipe.objects.count() == 1
+            recipe = Recipe.objects.get()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
         def test_it_can_create_recipes_actions_without_implementation(self, api_client):
             action = ActionFactory(implementation=None)
             assert action.implementation is None
@@ -437,6 +478,41 @@ class TestRecipeAPI(object):
             api_client.patch(f"/api/v2/recipe/{recipe.pk}/", {"name": "Test Recipe"})
             recipe.refresh_from_db()
             assert recipe.latest_revision.user is not None
+
+        def test_it_can_update_recipes_with_only_filter_object(self, api_client):
+            recipe = RecipeFactory(name="unchanged", extra_filter_expression="true")
+
+            res = api_client.patch(
+                "/api/v2/recipe/%s/" % recipe.id,
+                {
+                    "name": "changed",
+                    "extra_filter_expression": "",
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                },
+            )
+            assert res.status_code == 200, res.json()
+            recipe.refresh_from_db()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_object
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
+            # And you can omit it too
+            res = api_client.patch(
+                "/api/v2/recipe/%s/" % recipe.id,
+                {"name": "changed", "filter_object": [{"type": "channel", "channels": ["beta"]}]},
+            )
+            assert res.status_code == 200, res.json()
+            recipe.refresh_from_db()
+            assert recipe.extra_filter_expression == ""
+
+            # Let's paranoid-check that you can't unset the filter_object too.
+            res = api_client.patch(
+                "/api/v2/recipe/%s/" % recipe.id, {"name": "changed", "filter_object": []}
+            )
+            assert res.status_code == 400
+            assert res.json()["non_field_errors"] == [
+                "if extra_filter_expression is blank, at least one filter_object is required"
+            ]
 
     @pytest.mark.django_db
     class TestFilterObjects(object):

--- a/normandy/recipes/tests/api/v3/test_api.py
+++ b/normandy/recipes/tests/api/v3/test_api.py
@@ -371,7 +371,7 @@ class TestRecipeAPI(object):
             action = ActionFactory()
 
             res = api_client.post(
-                "/api/v2/recipe/",
+                "/api/v3/recipe/",
                 {
                     "name": "Test Recipe",
                     "action_id": action.id,
@@ -391,8 +391,25 @@ class TestRecipeAPI(object):
         def test_it_can_create_extra_filter_expression_omitted(self, api_client):
             action = ActionFactory()
 
+            # First try to create a recipe with 0 filter objects.
             res = api_client.post(
-                "/api/v2/recipe/",
+                "/api/v3/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "filter_object": [],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 400
+            assert res.json()["non_field_errors"] == [
+                "one of extra_filter_expression or filter_object is required"
+            ]
+
+            # Setting at least some filter_object but omitting the extra_filter_expression.
+            res = api_client.post(
+                "/api/v3/recipe/",
                 {
                     "name": "Test Recipe",
                     "action_id": action.id,

--- a/normandy/recipes/tests/api/v3/test_api.py
+++ b/normandy/recipes/tests/api/v3/test_api.py
@@ -367,6 +367,47 @@ class TestRecipeAPI(object):
             recipe = Recipe.objects.get()
             assert recipe.latest_revision.user is not None
 
+        def test_it_can_create_recipes_with_only_filter_object(self, api_client):
+            action = ActionFactory()
+
+            res = api_client.post(
+                "/api/v2/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "extra_filter_expression": "   ",
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 201, res.json()
+
+            assert Recipe.objects.count() == 1
+            recipe = Recipe.objects.get()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
+        def test_it_can_create_extra_filter_expression_omitted(self, api_client):
+            action = ActionFactory()
+
+            res = api_client.post(
+                "/api/v2/recipe/",
+                {
+                    "name": "Test Recipe",
+                    "action_id": action.id,
+                    "arguments": {},
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                    "enabled": True,
+                },
+            )
+            assert res.status_code == 201, res.json()
+
+            assert Recipe.objects.count() == 1
+            recipe = Recipe.objects.get()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
     @pytest.mark.django_db
     class TestUpdates(object):
         def test_it_can_edit_recipes(self, api_client):
@@ -472,6 +513,41 @@ class TestRecipeAPI(object):
             api_client.patch(f"/api/v3/recipe/{recipe.pk}/", {"name": "Test Recipe"})
             recipe.refresh_from_db()
             assert recipe.latest_revision.user is not None
+
+        def test_it_can_update_recipes_with_only_filter_object(self, api_client):
+            recipe = RecipeFactory(name="unchanged", extra_filter_expression="true")
+
+            res = api_client.patch(
+                "/api/v3/recipe/%s/" % recipe.id,
+                {
+                    "name": "changed",
+                    "extra_filter_expression": "",
+                    "filter_object": [{"type": "channel", "channels": ["beta"]}],
+                },
+            )
+            assert res.status_code == 200, res.json()
+            recipe.refresh_from_db()
+            assert recipe.extra_filter_expression == ""
+            assert recipe.filter_object
+            assert recipe.filter_expression == 'normandy.channel in ["beta"]'
+
+            # And you can omit it too
+            res = api_client.patch(
+                "/api/v3/recipe/%s/" % recipe.id,
+                {"name": "changed", "filter_object": [{"type": "channel", "channels": ["beta"]}]},
+            )
+            assert res.status_code == 200, res.json()
+            recipe.refresh_from_db()
+            assert recipe.extra_filter_expression == ""
+
+            # Let's paranoid-check that you can't unset the filter_object too.
+            res = api_client.patch(
+                "/api/v3/recipe/%s/" % recipe.id, {"name": "changed", "filter_object": []}
+            )
+            assert res.status_code == 400
+            assert res.json()["non_field_errors"] == [
+                "if extra_filter_expression is blank, at least one filter_object is required"
+            ]
 
     @pytest.mark.django_db
     class TestFilterObjects(object):

